ocfweb/docs/views/servers.py

```diff
@@ -57,81 +57,89 @@ def has_munin(self):
+            hosts[hostname] = (Host(hostname, h['type'], description, ()))
+    return hosts
+
+
+@periodic(300)
 def get_hosts():
-    return [
-        Host.from_ldap(
-            hostname='hal',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'fallingrocks',
-                    'monsoon',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='jaws',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'anthrax',
-                    'biohazard',
-                    'death',
-                    'dementors',
-                    'democracy',
-                    'firestorm',
-                    'flood',
-                    'lightning',
-                    'maelstrom',
-                    'pestilence',
-                    'reaper',
-                    'supernova',
-                    'thunder',
-                    'tsunami',
-                    'werewolves',
-                    'whirlwind',
-                    'whiteout',
-                    'zombies',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='pandemic',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'pileup',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='riptide',
-            type='hypervisor',
-        ),
-
-        Host.from_ldap(
-            hostname='corruption',
-            type='server',
-        ),
-
-        Host('blackhole', 'network', 'Managed Cisco Catalyst 2960S-48TS-L Switch.', []),
-
-        Host('logjam', 'printer', '', []),
-        Host('pagefault', 'printer', '', []),
-        Host('papercut', 'printer', '', []),
-        Host.from_ldap('overheat', type='raspi'),
-        Host.from_ldap('tornado', type='nuc'),
-
-        *(Host.from_ldap(desktop, type='desktop')
-          for desktop in sorted(list_desktops())),
-    ]
+    servers = create_hosts(hosts_by_filter('(type=server)'))
+    desktops = create_hosts(hosts_by_filter('(type=desktop)'))
+    misc = create_hosts(hosts_by_filter('(type=printer)'))
+    hypervisors = {}
+
+    hypervisor_hostnames = query_puppet(PQL_IS_HYPERVISOR)
+    all_children = query_puppet(PQL_GET_CHILDREN)
+
+    # Add children to hypervisors
+    for h in list(servers.values()):
+        if h.hostname in hypervisor_hostnames:
+            # Populate a list of children
+            children = []
+            for child_hostname in all_children.get(h.hostname, []):
+                child = servers.get(child_hostname)
+                if child:
+                    del servers[child.hostname]
+                    children.append(Host(child.hostname, 'vm', child.description, ()))
```

Comment:
You shouldn't need to construct a new `Host` object here, it should be fine to just append `child`.

Reply:
A new `Host` object is constructed for the sole purpose of changing its type from a physical `server` to `vm` given that they are not differentiated in LDAP listings.

Reply:
Instead of creating a new `Host` object, maybe I could call the `_replace` method?

Reply:
`_replace` will still return a new `Host` object (`namedtuple`s are immutable), but yes you should use that instead.